code/modules/lighting/lighting_turf.dm

```diff
@@ -1,26 +1,102 @@
 /turf
-	var/list/affecting_lights
-	var/atom/movable/lighting_overlay/lighting_overlay
+	var/dynamic_lighting = TRUE
+	luminosity           = 1
 
+	var/tmp/lighting_corners_initialised = FALSE
+
+	var/tmp/list/datum/light_source/affecting_lights       // List of light sources affecting this turf.
+	var/tmp/atom/movable/lighting_overlay/lighting_overlay // Our lighting overlay.
+	var/tmp/list/datum/lighting_corner/corners
+	var/tmp/has_opaque_atom = FALSE // Not to be confused with opacity, this will be TRUE if there's any opaque atom on the tile.
+
+/turf/New()
+	. = ..()
+
+	if(opacity)
+		has_opaque_atom = TRUE
+
+// Causes any affecting light sources to be queued for a visibility update, for example a door got opened.
 /turf/proc/reconsider_lights()
 	for(var/datum/light_source/L in affecting_lights)
 		L.vis_update()
 
-/turf/proc/lighting_clear_overlays()
+/turf/proc/lighting_clear_overlay()
 	if(lighting_overlay)
 		qdel(lighting_overlay)
 
-/turf/proc/lighting_build_overlays()
+	for(var/datum/lighting_corner/C in corners)
+		C.update_active()
+
+// Builds a lighting overlay for us, but only if our area is dynamic.
+/turf/proc/lighting_build_overlay()
 	if(lighting_overlay)
 		return
 
 	var/area/A = loc
-	if(A.lighting_use_dynamic)
-		var/atom/movable/lighting_overlay/O = new(src)
-		lighting_overlay = O
-		// forcibly refresh lighting on this tile
-		// must make light forget tile or it wil remain
-		// convinced that it is already lighting it
-		for(var/datum/light_source/L in affecting_lights)
-			L.forget_turf(src)
-			L.vis_update()
\ No newline at end of file
+	if(A.dynamic_lighting)
+		if(!lighting_corners_initialised)
+			generate_missing_corners()
+
+		new /atom/movable/lighting_overlay(src)
+
+		for(var/datum/lighting_corner/C in corners)
+			if(!C.active) // We would activate the corner, calculate the lighting for it.
+				for(var/L in C.affecting)
+					var/datum/light_source/S = L
+					S.recalc_corner(C)
+
+				C.active = TRUE
+
+// Used to get a scaled lumcount.
+/turf/proc/get_lumcount(var/minlum = 0, var/maxlum = 1)
+	if(!lighting_overlay)
+		return 0.5
```

Comment:
> /atom/movable/lighting_overlay/proc/get_clamped_lum has been removed, succeeded by /turf/proc/get_lumcount. They behave identically.

Really? This is supposed to return 1 for a turf without a lighting overlay, because a turf without one is fully lit.